// @ionic/cloud-angular/dist/es5/index.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
__export(require('@ionic/cloud'));
var rxjs_1 = require('rxjs');
var core_1 = require('@angular/core');
var cloud_2 = require('@ionic/cloud');
var Rx = (function () {
    function Rx(emitter) {
        this.emitter = emitter;
    }
    return Rx;
}());
exports.Rx = Rx;
var PushRx = (function (_super) {
    __extends(PushRx, _super);
    function PushRx() {
        _super.apply(this, arguments);
    }
    PushRx.prototype.notification = function () {
        var _this = this;
        return rxjs_1.Observable.fromEventPattern(function (h) {
            return _this.emitter.on('push:notification', function (data) {
                return h(data.message);
            });
        }, function (_) {
            // https://github.com/ReactiveX/rxjs/issues/1900
            // this.emitter.off(signal);
        });
    };
    return PushRx;
}(Rx));
exports.PushRx = PushRx;
var FacebookAuth = (function (_super) {
    __extends(FacebookAuth, _super);
    function FacebookAuth() {
        _super.apply(this, arguments);
    }
    FacebookAuth.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    FacebookAuth.ctorParameters = [];
    return FacebookAuth;
}(cloud_2.FacebookAuth));
exports.FacebookAuth = FacebookAuth;
var GoogleAuth = (function (_super) {
    __extends(GoogleAuth, _super);
    function GoogleAuth() {
        _super.apply(this, arguments);
    }
    GoogleAuth.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    GoogleAuth.ctorParameters = [];
    return GoogleAuth;
}(cloud_2.GoogleAuth));
exports.GoogleAuth = GoogleAuth;
var Auth = (function (_super) {
    __extends(Auth, _super);
    function Auth() {
        _super.apply(this, arguments);
    }
    Auth.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    Auth.ctorParameters = [];
    return Auth;
}(cloud_2.Auth));
exports.Auth = Auth;
var Client = (function (_super) {
    __extends(Client, _super);
    function Client() {
        _super.apply(this, arguments);
    }
    Client.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    Client.ctorParameters = [];
    return Client;
}(cloud_2.Client));
exports.Client = Client;
var Config = (function (_super) {
    __extends(Config, _super);
    function Config() {
        _super.apply(this, arguments);
    }
    Config.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    Config.ctorParameters = [];
    return Config;
}(cloud_2.Config));
exports.Config = Config;
var Deploy = (function (_super) {
    __extends(Deploy, _super);
    function Deploy() {
        _super.apply(this, arguments);
    }
    Deploy.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    Deploy.ctorParameters = [];
    return Deploy;
}(cloud_2.Deploy));
exports.Deploy = Deploy;
var DIContainer = (function (_super) {
    __extends(DIContainer, _super);
    function DIContainer() {
        _super.apply(this, arguments);
    }
    DIContainer.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    DIContainer.ctorParameters = [];
    return DIContainer;
}(cloud_2.DIContainer));
exports.DIContainer = DIContainer;
var Insights = (function (_super) {
    __extends(Insights, _super);
    function Insights() {
        _super.apply(this, arguments);
    }
    Insights.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    Insights.ctorParameters = [];
    return Insights;
}(cloud_2.Insights));
exports.Insights = Insights;
var Push = (function (_super) {
    __extends(Push, _super);
    function Push() {
        _super.apply(this, arguments);
    }
    Push.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    Push.ctorParameters = [];
    return Push;
}(cloud_2.Push));
exports.Push = Push;
var User = (function (_super) {
    __extends(User, _super);
    function User() {
        _super.apply(this, arguments);
    }
    User.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    User.ctorParameters = [];
    return User;
}(cloud_2.User));
exports.User = User;
exports.CloudSettingsToken = new core_1.OpaqueToken('CloudSettings');
function provideContainer(settings) {
    var container = new DIContainer();
    container.config.register(settings);
    container.core.init();
    container.cordova.bootstrap();
    return container;
}
exports.provideContainer = provideContainer;
function provideConfig(container) {
    return container.config;
}
exports.provideConfig = provideConfig;
function provideAuth(container) {
    return container.auth;
}
exports.provideAuth = provideAuth;
function provideClient(container) {
    return container.client;
}
exports.provideClient = provideClient;
function provideDeploy(container) {
    return container.deploy;
}
exports.provideDeploy = provideDeploy;
function provideUser(container) {
    return container.singleUserService.current();
}
exports.provideUser = provideUser;
function providePush(container) {
    var push = container.push;
    push.rx = new PushRx(container.eventEmitter);
    return push;
}
exports.providePush = providePush;
function provideFacebookAuth(container) {
    return container.facebookAuth;
}
exports.provideFacebookAuth = provideFacebookAuth;
function provideGoogleAuth(container) {
    return container.googleAuth;
}
exports.provideGoogleAuth = provideGoogleAuth;
var CloudModule = (function () {
    function CloudModule() {
    }
    CloudModule.forRoot = function (settings) {
        return {
            ngModule: CloudModule,
            providers: [
                { provide: exports.CloudSettingsToken, useValue: settings },
                { provide: DIContainer, useFactory: provideContainer, deps: [exports.CloudSettingsToken] },
                { provide: Auth, useFactory: provideAuth, deps: [DIContainer] },
                { provide: Client, useFactory: provideClient, deps: [DIContainer] },
                { provide: Config, useFactory: provideConfig, deps: [DIContainer] },
                { provide: Deploy, useFactory: provideDeploy, deps: [DIContainer] },
                { provide: Push, useFactory: providePush, deps: [DIContainer] },
                { provide: User, useFactory: provideUser, deps: [DIContainer] },
                { provide: FacebookAuth, useFactory: provideFacebookAuth, deps: [DIContainer] },
                { provide: GoogleAuth, useFactory: provideGoogleAuth, deps: [DIContainer] }
            ]
        };
    };
    CloudModule.decorators = [
        { type: core_1.NgModule },
    ];
    /** @nocollapse */
    CloudModule.ctorParameters = [];
    return CloudModule;
}());
exports.CloudModule = CloudModule;
